canvas/async_enroll_gatherer.py

```diff
@@ -0,0 +1,210 @@
+
+    def get_incomplete_course_ids(self) -> Sequence[int]:
+        course_ids = []
+
+        # Get unstarted course_ids
+        for course_id in self.course_ids:
+            if course_id not in self.course_enrollments.keys():
+                course_ids.append(course_id)
+
+        # Get in-progress course ids
+        for course_id in self.course_enrollments.keys():
+            course_enrollment_dict = self.course_enrollments[course_id]
+            if course_enrollment_dict['page_info']['hasNextPage']:
+                course_ids.append(course_id)
+
+        return course_ids
+
+    def parse_enrollment_response(self, future_response: Future) -> None:
+        # Check for irregular results
+        problem_encountered = False
+        response = future_response.result()
+
+        status_code = response.status_code
+        if status_code != 200:
+            logger.warning(f'Received irregular status code: {status_code}')
+            logger.debug(response.text)
+            problem_encountered = True
+        else:
+            try:
+                response_data = json.loads(response.text)
+            except JSONDecodeError:
+                logger.warning('JSONDecodeError encountered')
+                problem_encountered = True
+
+        if problem_encountered:
+            logger.warning('No data will be stored, and the request will be re-tried')
+        else:
+            response_course_id = int(response_data['data']['course']['_id'])
+
+            enrollments_connection = response_data['data']['course']['enrollmentsConnection']
+            enrollment_dicts = enrollments_connection['nodes']
+            enrollment_page_info = enrollments_connection['pageInfo']
+
+            if response_course_id not in self.course_enrollments.keys():
+                # Create new in-progress record
+                self.course_enrollments[response_course_id] = {
+                    'enrollments': enrollment_dicts,
+                    'page_info': enrollment_page_info,
+                    'num_pages': 1
+                }
+            else:
+                # Update existing in-progress record
+                self.course_enrollments[response_course_id]['enrollments'] += enrollment_dicts
+                self.course_enrollments[response_course_id]['page_info'] = enrollment_page_info
+                self.course_enrollments[response_course_id]['num_pages'] += 1
+
+    def make_requests(self, course_ids: Sequence[int]) -> None:
+        with FuturesSession(max_workers=self.num_workers) as session:
+            responses = []
+            for course_id in course_ids:
+                # Prep params
+                params = copy.deepcopy(self.default_params)
+                params['variables']['courseID'] = course_id
+                if course_id in self.course_enrollments.keys():
+                    course_enrollment_dict = self.course_enrollments[course_id]
+                    enroll_page_info = course_enrollment_dict['page_info']
```

Comment:
Will the case ever occur when a course already went through the loop and there was some 500 error and put on this retry list again be missed? I am guessing here once a loop is done it will go the `while` routine and further picked up for calling. is this how the logic happening?

Reply:
Yeah, so, if a request is made for a course and it fails (`problem_encountered`) because of an irregular status code or `JSONDecodeError`, then no record will be created in `self.course_enrollments`. When the loop starts over (which will happen once the first page is asked for from every course), it will try this course again. It will keep trying to ask for this course on each loop so long as there are uncompleted courses (courses with more pages). Does that help explain? There may be a better way to do this.